cpp/examples/arrow/execution_plan_documentation_examples.cc

```diff
@@ -0,0 +1,925 @@
+#include <arrow/dataset/plan.h>
+#include <arrow/dataset/scanner.h>
+
+#include <arrow/io/interfaces.h>
+#include <arrow/io/memory.h>
+
+#include <arrow/result.h>
+#include <arrow/status.h>
+#include <arrow/table.h>
+
+#include <arrow/ipc/api.h>
+
+#include <arrow/util/future.h>
+#include <arrow/util/range.h>
+#include <arrow/util/thread_pool.h>
+#include <arrow/util/vector.h>
+
+// Demonstrate various operators in Arrow Streaming Execution Engine
+
+namespace cp = ::arrow::compute;
+
+constexpr char kSep[] = "******";
+
+void PrintBlock(const std::string& msg) {
+  std::cout << "\n\t" << kSep << " " << msg << " " << kSep << "\n" << std::endl;
+}
+
+template <typename TYPE,
+          typename = typename std::enable_if<arrow::is_number_type<TYPE>::value |
+                                             arrow::is_boolean_type<TYPE>::value |
+                                             arrow::is_temporal_type<TYPE>::value>::type>
+arrow::Result<std::shared_ptr<arrow::Array>> GetArrayDataSample(
+    const std::vector<typename TYPE::c_type>& values) {
+  using ARROW_ARRAY_TYPE = typename arrow::TypeTraits<TYPE>::ArrayType;
+  using ARROW_BUILDER_TYPE = typename arrow::TypeTraits<TYPE>::BuilderType;
+  ARROW_BUILDER_TYPE builder;
+  ABORT_NOT_OK(builder.Reserve(values.size()));
+  std::shared_ptr<ARROW_ARRAY_TYPE> array;
+  ABORT_NOT_OK(builder.AppendValues(values));
+  ABORT_NOT_OK(builder.Finish(&array));
+  arrow::Result<std::shared_ptr<ARROW_ARRAY_TYPE>> result(std::move(array));
+  return result;
+}
+
+template <class TYPE>
+arrow::Result<std::shared_ptr<arrow::Array>> GetBinaryArrayDataSample(
+    const std::vector<std::string>& values) {
+  using ARROW_ARRAY_TYPE = typename arrow::TypeTraits<TYPE>::ArrayType;
+  using ARROW_BUILDER_TYPE = typename arrow::TypeTraits<TYPE>::BuilderType;
+  ARROW_BUILDER_TYPE builder;
+  ABORT_NOT_OK(builder.Reserve(values.size()));
+  std::shared_ptr<ARROW_ARRAY_TYPE> array;
+  ABORT_NOT_OK(builder.AppendValues(values));
+  ABORT_NOT_OK(builder.Finish(&array));
+  arrow::Result<std::shared_ptr<ARROW_ARRAY_TYPE>> result(std::move(array));
+  return result;
+}
+
+arrow::Result<std::shared_ptr<arrow::RecordBatch>> GetSampleRecordBatch(
+    const arrow::ArrayVector array_vector, const arrow::FieldVector& field_vector) {
+  std::shared_ptr<arrow::RecordBatch> record_batch;
+  ARROW_ASSIGN_OR_RAISE(auto struct_result,
+                        arrow::StructArray::Make(array_vector, field_vector));
+  return record_batch->FromStructArray(struct_result);
+}
+
+/**
+ * \brief Get the Dataset object
+ *  Creating Dataset
+ *  a, b
+    1,null
+    2,true
+    null,true
+    3,false
+    null,true
+    4,false
+    5,null
+    6,false
+    7,false
+    8,true
+ * \return arrow::Result<std::shared_ptr<arrow::dataset::Dataset>>
+ */
+arrow::Result<std::shared_ptr<arrow::dataset::Dataset>> GetDataset() {
+  auto null_long = std::numeric_limits<int64_t>::quiet_NaN();
+  auto null_bool = std::numeric_limits<bool>::quiet_NaN();
```

Comment:
Is this valid? `bool` has no NaN value.

Reply:
We probably need a separate overload that takes both an array of values and an array of validity bits.